pkg/machinehealthcheck/reconciler.go

```diff
@@ -0,0 +1,109 @@
+)
+
+// Reconciler allows to reconcile machine health checks.
+type Reconciler struct {
+	client client.Client
+}
+
+// New returns a new Reconciler.
+func New(client client.Client) *Reconciler {
+	return &Reconciler{
+		client: client,
+	}
+}
+
+// InstallMachineHealthChecks installs machine health checks for a given cluster.
+func (r *Reconciler) InstallMachineHealthChecks(ctx context.Context, log logr.Logger, cluster *anywherev1.Cluster) error {
+	log.Info("Installing machine health checks for workload", "cluster", cluster.Name)
+
+	if cluster.Spec.MachineHealthCheck != nil {
+		setMachineHealthCheckTimeoutDefaults(cluster)
+	} else {
+
+		// check if machine health checks have already been installed on the cluster and populate the EKSA spec
+		mhc, err := r.machineHealthChecksExists(ctx, log, cluster)
+		if err != nil {
+			return err
+		}
+		if mhc != nil {
+			cluster.Spec.MachineHealthCheck = &anywherev1.MachineHealthCheck{
+				NodeStartupTimeout:      mhc.Spec.NodeStartupTimeout.OpenAPISchemaFormat(),
+				UnhealthyMachineTimeout: mhc.Spec.UnhealthyConditions[0].Timeout.String(),
+			}
+		} else {
+			cluster.Spec.MachineHealthCheck = &anywherev1.MachineHealthCheck{}
+			setMachineHealthCheckTimeoutDefaults(cluster)
+		}
+	}
+
+	unhealthyMachineTimeout, err := time.ParseDuration(cluster.Spec.MachineHealthCheck.UnhealthyMachineTimeout)
+	if err != nil {
+		return err
+	}
+
+	nodeStartupTimeout, err := time.ParseDuration(cluster.Spec.MachineHealthCheck.NodeStartupTimeout)
+	if err != nil {
+		return err
+	}
+
+	mhc := clusterapi.MachineHealthCheckObjects(cluster, unhealthyMachineTimeout, nodeStartupTimeout)
+
+	err = serverside.ReconcileObjects(ctx, r.client, clientutil.ObjectsToClientObjects(mhc))
+	if err != nil {
+		return fmt.Errorf("applying machine health checks: %v", err)
+	}
+
+	return nil
+}
+
+// machineHealthChecksExists checks if machine health checks already exist on the cluster and returns it.
+func (r *Reconciler) machineHealthChecksExists(ctx context.Context, log logr.Logger, cluster *anywherev1.Cluster) (*clusterv1.MachineHealthCheck, error) {
+	mhc := &clusterv1.MachineHealthCheck{}
+	// we get only kcp machine health check as it had the same timeouts as worker machine health checks
+	kcpMHCName := clusterapi.ControlPlaneMachineHealthCheckName(cluster)
+	err := r.client.Get(ctx, types.NamespacedName{Name: kcpMHCName, Namespace: constants.EksaSystemNamespace}, mhc)
+	if apierrors.IsNotFound(err) {
+		log.Info("Machine health checks don't exist for", "cluster", cluster.Name)
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "fetching machine health check %s", kcpMHCName)
+	}
+
+	return mhc, nil
+}
+
+// setMachineHealthCheckTimeoutDefaults sets default tiemout values for cluster's machine health checks.
+func setMachineHealthCheckTimeoutDefaults(cluster *anywherev1.Cluster) {
+	if len(cluster.Spec.MachineHealthCheck.NodeStartupTimeout) == 0 {
```

Comment:
This is another place we are running the defaults? I saw some defaulting above, but not sure if that was for only CLI or something.

Reply:
yes the cluster defaulter only sets defaults for the CLI. the ones in here set it for the controller.

Reply:
Can we combine this defaulting logic in one place instead of having it defined in before? That is in case we decide to change the default values in the future?